### PR TITLE
Fix analyzer treatment of flow captures of arrays

### PR DESCRIPTION
**Pull request by sbomer**
https://github.com/dotnet/runtime/pull/93259 uncovered an issue around how the analyzer tracks arrays. It hit an analyzer assert while trying to create an l-value flow capture of another flow capture reference, which should never happen as far as I understand. The assert was firing for https://github.com/dotnet/runtime/blob/946c5245aea149d9ece53fd0debc18328c29083b/src/libraries/System.Private.CoreLib/src/System/IO/RandomAccess.Windows.cs#L511
Now tested in TestNullCoalescingAssignment:
```csharp
(arr ??= arr2)[0] = typeof (V);
```
The CFG had three flow captures:
- capture 0: an l-value flow capture of `arr`. Used later in the branch that assigns `arr2` to `arr`.
- capture 1: an r-value flow capture of capture 0. This was checked for null.
- capture 2: an l-value flow capture representing `arr ??= arr2`, used to write index 0 of the array.
  - In the == null branch, this captured the result of an assignment (capture 0 = `arr2`)
  - In the other branch, it captured "capture 1". This is where the assert was hit.

```mermaid
flowchart TD
style 0 text-align: left;
0("<code></code>")
style 1 text-align: left;
1("<code>&nbsp;&nbsp;LocalReferenceOperation: arr = new Type[1] { typeof (T) }
&nbsp;&nbsp;&nbsp;&nbsp;LiteralOperation: 1
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;TypeOfOperation: typeof (T)
&nbsp;&nbsp;&nbsp;&nbsp;ArrayInitializerOperation: { typeof (T) }
&nbsp;&nbsp;ArrayCreationOperation: new Type[1] { typeof (T) }
SimpleAssignmentOperation: arr = new Type[1] { typeof (T) }
&nbsp;&nbsp;LocalReferenceOperation: arr2 = new Type[1] { typeof (U) }
&nbsp;&nbsp;&nbsp;&nbsp;LiteralOperation: 1
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;TypeOfOperation: typeof (U)
&nbsp;&nbsp;&nbsp;&nbsp;ArrayInitializerOperation: { typeof (U) }
&nbsp;&nbsp;ArrayCreationOperation: new Type[1] { typeof (U) }
SimpleAssignmentOperation: arr2 = new Type[1] { typeof (U) }
</code>")
0 --> 1
style 2 text-align: left;
2("<code>&nbsp;&nbsp;LocalReferenceOperation: arr
FlowCaptureOperation: arr (0)
</code>")
1 --> 2
style 3 text-align: left;
3("<code>&nbsp;&nbsp;FlowCaptureReferenceOperation: arr (0)
FlowCaptureOperation: arr (1)
&nbsp;&nbsp;FlowCaptureReferenceOperation: arr (1)
IsNullOperation: arr
</code>")
2 --> 3
style 4 text-align: left;
4("<code>&nbsp;&nbsp;FlowCaptureReferenceOperation: arr (1)
FlowCaptureOperation: arr ??= arr2 (2)
</code>")
3 --> 4
style 5 text-align: left;
5("<code>&nbsp;&nbsp;&nbsp;&nbsp;FlowCaptureReferenceOperation: arr (0)
&nbsp;&nbsp;&nbsp;&nbsp;LocalReferenceOperation: arr2
&nbsp;&nbsp;SimpleAssignmentOperation: arr ??= arr2
FlowCaptureOperation: arr ??= arr2 (2)
</code>")
3 --> 5
style 6 text-align: left;
6("<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;FlowCaptureReferenceOperation: arr ??= arr2 (2)
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;LiteralOperation: 0
&nbsp;&nbsp;&nbsp;&nbsp;ArrayElementReferenceOperation: (arr ??= arr2)[0]
&nbsp;&nbsp;&nbsp;&nbsp;TypeOfOperation: typeof (V)
&nbsp;&nbsp;SimpleAssignmentOperation: (arr ??= arr2)[0] = typeof (V)
ExpressionStatementOperation: (arr ??= arr2)[0] = typeof (V);
</code>")
4 --> 6
5 --> 6
style 7 text-align: left;
7("<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;LocalReferenceOperation: arr
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;LiteralOperation: 0
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;ArrayElementReferenceOperation: arr[0]
&nbsp;&nbsp;&nbsp;&nbsp;ArgumentOperation: arr[0]
&nbsp;&nbsp;InvocationOperation: arr[0].RequiresAll ()
ExpressionStatementOperation: arr[0].RequiresAll ();
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;LocalReferenceOperation: arr2
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;LiteralOperation: 0
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;ArrayElementReferenceOperation: arr2[0]
&nbsp;&nbsp;&nbsp;&nbsp;ArgumentOperation: arr2[0]
&nbsp;&nbsp;InvocationOperation: arr2[0].RequiresPublicFields ()
ExpressionStatementOperation: arr2[0].RequiresPublicFields ();
</code>")
6 --> 7
style 8 text-align: left;
8("<code></code>")
7 --> 8
```

The bug, I believe, is that capture 2 should have been an r-value flow capture instead. Even though it's used for writing to the array, the assignment doesn't modify the array pointer represented by this capture - it dereferences this pointer and modifies the array. This was introduced by my modifications to the l-value detection logic in https://github.com/dotnet/runtime/pull/90287.

This undoes that portion of the change so that capture 2 is now treated as an r-value capture. This simplifies the array element assignment logic so that it no longer can see an assignment where the array is an l-value.

Fixes https://github.com/dotnet/runtime/pull/93420 by adding an explicit check for `IsInitialization` so that we don't hit the related asserts for string interpolation handlers.

### CODE REVIEW

File: src/tools/illink/src/ILLink.RoslynAnalyzer/DataFlow/LocalDataFlowVisitor.cs [hunk 1]
```diff
@@ -371,19 +337,10 @@ public override TValue VisitSimpleAssignment (ISimpleAssignmentOperation operati
...
 		TValue GetFlowCaptureValue (IFlowCaptureReferenceOperation operation, LocalDataFlowState<TValue, TValueLattice> state)
 		{
-			if (!operation.GetValueUsageInfo (OwningSymbol).HasFlag (ValueUsageInfo.Read)) {
-				// There are known cases where this assert doesn't hold, because LValueFlowCaptureProvider
-				// produces the wrong result in some cases for flow captures with IsInitialization = true.
-				// https://github.com/dotnet/linker/issues/2749
-				// Debug.Assert (IsLValueFlowCapture (operation.Id));
+			Debug.Assert (!IsLValueFlowCapture (operation.Id));
+			Debug.Assert (operation.GetValueUsageInfo (OwningSymbol).HasFlag (ValueUsageInfo.Read));
+			if (!operation.GetValueUsageInfo (OwningSymbol).HasFlag (ValueUsageInfo.Read))
```
Comment (vitek-karas) [inline src/tools/illink/src/ILLink.RoslynAnalyzer/DataFlow/LocalDataFlowVisitor.cs:342]: Does this resolve https://github.com/dotnet/linker/issues/2749?

**Review by vitek-karas — COMMENTED** [review]
I must admit this is a bit too much compiler theory for me :-)

I think it does make sense, but it would be really good if @agocke could also take a look at this.

**Comment by agocke** [comment]
Looking....

Also, cool Mermaid graph :)

File: src/tools/illink/src/ILLink.RoslynAnalyzer/DataFlow/LocalDataFlowVisitor.cs [hunk 2]
```diff
@@ -235,45 +235,11 @@ TValue ProcessSingleTargetAssignment (IOperation targetOperation, ISimpleAssignm
 					if (arrayElementRef.Indices.Length != 1)
 						break;
 
-					// Similarly to VisitSimpleAssignment, this needs to handle cases where the array reference
```
[the remark this answers is not in the document]
Comment (agocke) [reply src/tools/illink/src/ILLink.RoslynAnalyzer/DataFlow/LocalDataFlowVisitor.cs:238]: By this you mean, that the array (LHS) part of the ArrayReference should be an rvalue, right? Not the thing as a whole?

**Review by agocke — APPROVED** [review]
This LGTM, but I can't promise I'm not missing something. Abstractly, "lifting into a temporary" seems like it can encompass a lot of things. I'm not sure I've thought of all the cases.

**Pull request merged**